backend/app/anti_recommenders/arkg/arkg_anti_recommender.py

```diff
@@ -1,162 +1,209 @@
+                if unseen_anti_recommendation_key
+                not in self.__user_state.anti_recommendations_history.values()
+            ]
+
+            if not primary_anti_recommendation_keys:
+                primary_anti_recommendation_keys = [
+                    unseen_anti_recommendation_key
+                    for unseen_anti_recommendation_key in self.__record_keys
+                    if unseen_anti_recommendation_key
+                    not in self.__user_state.anti_recommendations_history.values()
+                ]
+
+            if primary_anti_recommendation_keys:
+                return primary_anti_recommendation_keys[0]
 
-        return tuple(anti_recommendation_path_graph_list)
+        return None
 
-    @typing.override
+    @override
     def generate_anti_recommendations(
         self, *, record_key: RecordKey
     ) -> Iterable[AntiRecommendation]:
         """
-        Generate anti-recommendations of a Record.
+        Generate anti-recommendations of a `Record`.
 
-        Anti-recommendations are obtained from an ARKG.
+        Anti-recommendations are obtained from an `ARKG`.
+
+        The first anti-recommendation generated is an anti-recommendation that has not yet been seen by a `User`.
+
+        All other anti-recommendations may or may not have been seen by a User.
         """
 
-        if not self.__anti_recommendation_path_graph:
-            self.__anti_recommendation_path_graph = list(
-                self.__build_anti_recommendation_path_graph(record_key)
+        if self.__user_state:
+            anti_recommendation_keys = list(
+                self.__retrieve_anti_recommendations_from_store(record_key)
+            )
+
+            primary_anti_recommendation_key = (
+                self.__select_primary_anti_recommendation_key(
+                    tuple(anti_recommendation_keys)
+                )
             )
 
-        return self.__anti_recommendation_path_graph.pop().anti_recommendations
+            if primary_anti_recommendation_key:
+                if primary_anti_recommendation_key in anti_recommendation_keys:
+                    anti_recommendation_keys.remove(primary_anti_recommendation_key)
+
+                anti_recommendation_keys.insert(0, primary_anti_recommendation_key)
+
+                self.__user_state.anti_recommendations_history.update(
+                    {"anti_recommendation_key": primary_anti_recommendation_key}
+                )
+
+                try:
+                    self.__upsert_user_state_into_database()
+
+                    return (
+                        AntiRecommendation(
+                            key=anti_recommendation_key,
+                            url=WIKIPEDIA_BASE_URL + anti_recommendation_key,
+                        )
+                        for anti_recommendation_key in anti_recommendation_keys
+                    )
+                except APIError as exception:
```

Comment:
Per comment above, it feels very odd to have this sort of state management in the same class as business logic.